### PR TITLE
sql: implement pg_cancel_backend

### LINKED ISSUE #13129: sql: support `pg_cancel_backend`
### Feature request

The `pg_cancel_backend(pid)` function cancels the query that is currently running on identified connection, if any. It is used by some PostgreSQL tools. It is equivalent in behavior to the pgwire-level cancellation signal, which we already support.

https://www.postgresql.org/docs/current/functions-admin.html#FUNCTIONS-ADMIN-SIGNAL

### LINKED ISSUE #19773: ws: support query cancellation
*Part of epic #19771.*

It's not (easily?) possible today to cancel a query running on a WebSocket connection. This is particularly annoying for `SUBSCRIBE` queries, which typically are not expected to terminate until they are explicitly canceled.

From https://github.com/MaterializeInc/materialize/issues/18224#issuecomment-1562286516:

> I actually think you _could_ cancel the subscribe if you could somehow figure out the connection ID for the session and then issue a cancellation request via pgwire. But we should probably support this natively in the WS protocol. Probably the best thing to do is to allow the client to send a `{"cancel": true}` message at any point to cancel the in-flight command on the connection.
> 
> Don't think we want to follow the pgwire approach of having a separate connection send a cancel request. The justification for why it works like this in pgwire seems stuck in 1998.
> 
> > During the processing of a query, the frontend might request cancellation of the query. The cancel request is not sent directly on the open connection to the backend for reasons of implementation efficiency: we don't want to have the backend constantly checking for new input from the frontend during query processing. Cancel requests should be relatively infrequent, so we make them slightly cumbersome in order to avoid a penalty in the normal case.
> 
> We have access to async/await in Rust, so it is perfectly efficient to both stream query results to the client while checking to see if the client has sent a cancel request.

The ask here is roughly to allow the WS client to send a message like

```
{"cancel": true}
```

at any point, and have the server issue a cancellation request for whatever is currently running on the connection.

### PR DESCRIPTION
**Pull request by mjibson**
Parameters don't work on it yet, but it's good enough to use with a hard coded parameter.

Fixes #19773.
Fixes #13129.

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Add `pg_cancel_backend()` to cancel queries from SQL.

### CODE REVIEW

File: src/adapter/src/coord.rs
Comment (jkosh44) [inline src/adapter/src/coord.rs:478]: I looked through the PostgreSQL to see how they handle this. They actually don't use the current role or the session role. They use the authenticated role. The authenticated role is the role that the current session authenticated with at connection time and it never changes for the entire duration of the session.

I haven't added the authenticated role to Materialize yet unfortunately. However, you can remove the comment here about updating the role because it should never get updated. Can you also update the first part to indicate that it's the authenticated role and not the current role.

In case you're interested in the PG implementation, I've included some links below:

Here is where `pg_cancel_backend` is implemented, which mostly delegates to `pg_signal_backend`:
https://github.com/postgres/postgres/blob/a14e75eb0b6a73821e0d66c0d407372ec8376105/src/backend/storage/ipc/signalfuncs.c#L110-L136

Here is the beginning of `pg_signal_backend`, where the metadata for the other process is retrieved and the permission check is done:
https://github.com/postgres/postgres/blob/a14e75eb0b6a73821e0d66c0d407372ec8376105/src/backend/storage/ipc/signalfuncs.c#L30-L84

Here is when a session's role in `static PGPROC *allProcs` get's initialized: 
https://github.com/postgres/postgres/blob/a14e75eb0b6a73821e0d66c0d407372ec8376105/src/backend/utils/init/miscinit.c#L778-L780

File: src/adapter/src/coord/command_handler.rs
Comment (jkosh44) [inline src/adapter/src/coord/command_handler.rs:283]: We should probably use the session role since it's closer to the authenticated role and leave a TODO saying switch this to the authenticated role when it's implemented.

File: src/environmentd/tests/sql.rs
Comment (jkosh44) [inline src/environmentd/tests/sql.rs:3288]: Which cast are you referring to?

File: src/sql/src/plan/statement/dml.rs
Comment (jkosh44) [inline src/sql/src/plan/statement/dml.rs:239]: Does `eval_expr` return an error if the arg isn't an `Int32`?

**Review by jkosh44 — APPROVED** [review]
Couple small notes, otherwise LGTM

File: src/sql/src/plan/statement/dml.rs
Comment (benesch) [inline src/sql/src/plan/statement/dml.rs:184]: I think you just want the `*` operator here? But I didn't actually try this.

```suggestion
        } if ctes.is_empty() && order_by.is_empty() => match *body {
```

File: src/sql/src/plan/statement/dml.rs
Comment (benesch) [inline src/sql/src/plan/statement/dml.rs:216]: I don't think `CastContext::Assignment` is right for function calls. I think you want `CastContext::Implicit`.

File: src/sql/src/plan/statement/dml.rs
Comment (benesch) [inline src/sql/src/plan/statement/dml.rs:210]:
```suggestion
                        fn eval_function_argument<'a>(
```

(Because the choice of `CastContext` is specific for function arguments.)

File: src/sql/src/plan/statement/dml.rs
Comment (benesch) [inline src/sql/src/plan/statement/dml.rs:227]: This is a user visible name for use in error messages. I think you want to match what we call this in `plan_view_select`, which is "SELECT clause."

```suggestion
                            name: "SELECT clause",
```

File: src/sql/src/func.rs
Comment (benesch) [inline src/sql/src/func.rs:2306]: Having this error message is great. Otherwise we'd gaslight users about whether this function is real or not, depending on how they type in their SQL queries.

File: src/adapter/src/coord.rs
Comment (benesch) [reply src/adapter/src/coord.rs:478]: Whoa, awesome sleuthing, @jkosh44!

File: src/sql/src/plan/statement/dml.rs
Comment (benesch) [reply src/sql/src/plan/statement/dml.rs:239]: Yes, the call to `cast_to` will error if the cast is impossible!

File: src/sql/src/plan.rs
Comment (benesch) [inline src/sql/src/plan.rs:1263]:
```suggestion
/// See `plan_select_if_side_effecting` for details.
```

File: src/sql/src/plan/statement/dml.rs
Comment (benesch) [inline src/sql/src/plan/statement/dml.rs:172]:
```suggestion
/// Plans the `SELECT` if it contains calls to side-effecting functions.
///
/// In PostgreSQL, these functions can appear anywhere in a query:
///
///     SELECT 1 WHERE pg_cancel_backend(1234)
///
/// In Materialize, our compute layer cannot execute functions with side effects. So we
/// sniff out the common form of function calls, i.e.
///
///    SELECT side_effecting_function(...)
///
/// where all arguments are literals or bound parameters, and plan them specially as a
/// `Plan::SideEffectingFunction`. This gets us compatibility with PostgreSQL for 90% of
/// real-world use cases, without causing stress for the compute layer (optimizer,
/// dataflow execution, etc.), as we can handle all the side effects cleanly in the adapter layer.
fn plan_select_if_side_effecting(
    // This match looks unwieldy, but it should be robust in practice. We exhaustively
    // match everywhere, so the compiler will warn us if we add new fields to the AST
    // that need to be handled here. 
```

(The `is_` naming makes me think it's just checking, when in fact it is also planning if it's a single side-ffecting function.)

File: src/sql/src/plan/statement/dml.rs
Comment (benesch) [inline src/sql/src/plan/statement/dml.rs:209]: You could save yourself a bunch of nesting with two `let ... else { return Ok(None) }`, if you like!

File: src/sql/src/plan/statement/dml.rs
Comment (benesch) [inline src/sql/src/plan/statement/dml.rs:263]:
```suggestion
    if let Some(f) = plan_select_if_side_effecting(scx, &stmt)? {
```

File: src/sql/src/plan/statement/dml.rs
Comment (benesch) [inline src/sql/src/plan/statement/dml.rs:278]:
```suggestion
    if let Some(f) = plan_select_if_side_effecting(scx, &select)? {
```

**Review by benesch — COMMENTED** [review]
Hallelujah! Thank you so much for finding a path forward for this function. Dropped in some comments but they are all minor; don't wait for my approval!

**Comment by benesch** [comment]
I pushed up a revision to support parameters. The good news is `pg_cancel_backend` now works with parameters. The bad news is that it was a large yak shave and needs a review. I'd suggest reviewing `side_effecting_func.rs` from scratch; everything else is unchanged. You can also skip over the first half of the `extract_sef_call` function that's walking the AST; that's basically unchanged from @mjibson's original code, just refactored from a nested match into `let ... else { return }` calls.

(I addressed all my own review feedback as well. Feel free to change anything back if any of my comments were controversial.)

**Review by benesch — APPROVED** [review]
Approving the parts of this that I didn't write!

File: src/adapter/src/coord/command_handler.rs
Comment (mjibson) [reply src/adapter/src/coord/command_handler.rs:283]: How does authenticated role differ from session role? Aren't those both set once at connection start?

Comment (jkosh44) [reply src/adapter/src/coord/command_handler.rs]: The session role can be modified via `SET SESSION AUTHORIZATION` (https://www.postgresql.org/docs/15/sql-set-session-authorization.html). The authenticated role will never change.

Comment (mjibson) [reply src/adapter/src/coord/command_handler.rs]: Ohh I didn't know this was a thing. So there's both `SET ROLE` and `SET SESSION AUTHORIZATION`? Neat?

Comment (jkosh44) [reply src/adapter/src/coord/command_handler.rs]: The authenticated role is used to determine privileges for:
- `SET SESSION AUTHORIZATION`.
- system admin functions, like `pg_cancel_backend`.
- Maybe others that I don't know about.

The session role is used to determine privileges for:
- `SET ROLE`/`RESET ROLE`.
- Maybe others that I don't know about.

**Pull request merged**
